classes/migration/upgrade/v3_5_0/I8333_AddMissingForeignKeys.php

```diff
@@ -0,0 +1,253 @@
+<?php
+
+/**
+ * @file classes/migration/upgrade/v3_5_0/I8333_AddMissingForeignKeys.php
+ *
+ * Copyright (c) 2023 Simon Fraser University
+ * Copyright (c) 2023 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class I8333_AddMissingForeignKeys
+ *
+ * @brief Upgrade/downgrade operations for introducing foreign key definitions to existing database relationships.
+ */
+
+namespace PKP\migration\upgrade\v3_5_0;
+
+use Illuminate\Database\Query\Builder;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use PKP\install\DowngradeNotSupportedException;
+
+abstract class I8333_AddMissingForeignKeys extends \PKP\migration\Migration
+{
+    abstract protected function getContextTable(): string;
+    abstract protected function getContextSettingsTable(): string;
+    abstract protected function getContextKeyField(): string;
+
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        $this->setupFieldSchema();
+        $this->updateSpecialValues();
+        $this->clearOrphanedEntities();
+        $this->setupForeignKeys();
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        throw new DowngradeNotSupportedException();
+    }
+
+    protected function updateSpecialValues(): void
+    {
+        DB::table('site')
+            ->where('redirect_context_id', '=', 0)
+            ->update(['redirect_context_id' => null]);
+        DB::table('filters')
+            ->where('parent_filter_id', '=', 0)
+            ->update(['parent_filter_id' => null]);
+        DB::table('navigation_menu_item_assignments')
+            ->where('parent_id', '=', 0)
+            ->update(['parent_id' => null]);
+        DB::table('email_log')
+            ->where('sender_id', '=', 0)
+            ->update(['sender_id' => null]);
+        // Only the administrator user group (role_id = 1) is allowed to have a null context_id
+        DB::table('user_groups')
+            ->where('role_id', '=', 1)
+            ->where('context_id', '=', 0)
+            ->update(['context_id' => null]);
```

Comment:
I would also suggest deleting any `user_groups` entries where `context_id > 0 AND role_id=1`. Entries like that might be an outcome of https://github.com/pkp/pkp-lib/issues/10180.

Reply:
I'll do it, it makes sense.